File: js/integration/_support/matrix.ts
export type MatrixCase = {
  readonly id: string;
  readonly fixture: string;
  readonly title: string;
  readonly coverage: readonly string[];
  readonly description: string;
};

export type ClientTestMatrix = {
  readonly schemaVersion: 1;
  readonly cases: readonly MatrixCase[];
};

const MATRIX_URL = new URL(
  "../../../integration/client-test-matrix.json",
  import.meta.url,
);

/** Loads and validates the shared client integration matrix. */
export async function loadClientTestMatrix(): Promise<ClientTestMatrix> {
  const text = await Deno.readTextFile(MATRIX_URL);
  const parsed: unknown = JSON.parse(text);
  return parseClientTestMatrix(parsed);
}

/** Returns sorted matrix case IDs. */
export function matrixCaseIds(matrix: ClientTestMatrix): string[] {
  return matrix.cases.map((caseEntry) => caseEntry.id).toSorted();
}

/** Returns matrix cases grouped by fixture name. */
export function matrixCasesByFixture(
  matrix: ClientTestMatrix,
): ReadonlyMap<string, readonly MatrixCase[]> {
  const groups = new Map<string, MatrixCase[]>();
  for (const caseEntry of matrix.cases) {
    const group = groups.get(caseEntry.fixture) ?? [];
    group.push(caseEntry);
    groups.set(caseEntry.fixture, group);
  }
  return groups;
}

function parseClientTestMatrix(value: unknown): ClientTestMatrix {
  const root = expectRecord(value, "matrix root");
  expectKeys(root, ["schemaVersion", "cases"], "matrix root");

  if (root.schemaVersion !== 1) {
    throw new Error("client integration matrix schemaVersion must be 1");
  }
  if (!Array.isArray(root.cases)) {
    throw new Error("client integration matrix cases must be an array");
  }

  const cases = root.cases.map(parseMatrixCase);
  const duplicateIds = duplicates(cases.map((caseEntry) => caseEntry.id));
  if (duplicateIds.length > 0) {
    throw new Error(
      `client integration matrix has duplicate case ids: ${
        duplicateIds.join(
          ", ",
        )
      }`,
    );
  }

  return { schemaVersion: 1, cases };
}

function parseMatrixCase(value: unknown, index: number): MatrixCase {
  const context = `matrix case ${index + 1}`;
  const caseEntry = expectRecord(value, context);
  expectKeys(
    caseEntry,
    ["id", "fixture", "title", "coverage", "description"],
    context,
  );

  const id = expectNonEmptyString(caseEntry.id, `${context} id`);
  const fixture = expectNonEmptyString(caseEntry.fixture, `${context} fixture`);
  if (!id.startsWith(`${fixture}.`)) {
    throw new Error(
      `${context} id ${id} must start with fixture prefix ${fixture}.`,
    );
  }
  if (!Array.isArray(caseEntry.coverage)) {
    throw new Error(`${context} coverage must be an array`);
  }
  const coverage = caseEntry.coverage.map((entry, coverageIndex) =>
    expectNonEmptyString(entry, `${context} coverage ${coverageIndex + 1}`)
  );

  return {
    id,
    fixture,
    title: expectNonEmptyString(caseEntry.title, `${context} title`),
    coverage,
    description: expectNonEmptyString(
      caseEntry.description,
      `${context} description`,
    ),
  };
}

function expectRecord(
  value: unknown,
  context: string,
): Record<string, unknown> {
  if (!isRecord(value)) {
    throw new Error(`${context} must be an object`);
  }
  return value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function expectKeys(
  value: Record<string, unknown>,
  allowedKeys: readonly string[],
  context: string,
): void {
  const allowed = new Set(allowedKeys);
  const unknownKeys = Object.keys(value).filter((key) => !allowed.has(key));
  if (unknownKeys.length > 0) {
    throw new Error(`${context} has unknown keys: ${unknownKeys.join(", ")}`);
  }
}

function expectNonEmptyString(value: unknown, context: string): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error(`${context} must be a non-empty string`);
  }
  return value;
}

function duplicates(values: readonly string[]): string[] {
  const seen = new Set<string>();
  const duplicateValues = new Set<string>();
  for (const value of values) {
    if (seen.has(value)) {
      duplicateValues.add(value);
    }
    seen.add(value);
  }
  return [...duplicateValues].toSorted();
}

File: rust/crates/trellis/tests/integration/support/matrix.rs
use std::collections::BTreeSet;
use std::fmt;
use std::fs;
use std::path::PathBuf;

use serde::de;
use serde::Deserialize;

#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) struct ClientTestMatrix {
    pub(crate) cases: Vec<MatrixCase>,
}

impl ClientTestMatrix {
    pub(crate) fn case_by_id(&self, id: &str) -> Option<&MatrixCase> {
        self.cases.iter().find(|case_entry| case_entry.id == id)
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) struct MatrixCase {
    pub(crate) id: String,
    pub(crate) fixture: String,
    pub(crate) title: String,
    pub(crate) coverage: Vec<String>,
    pub(crate) description: String,
    pub(crate) scenario: Scenario,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) struct Scenario {
    pub(crate) participants: Vec<ScenarioParticipant>,
    pub(crate) given: Vec<String>,
    pub(crate) when: Vec<String>,
    pub(crate) then: Vec<String>,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) struct ScenarioParticipant {
    pub(crate) name: String,
    pub(crate) kind: ScenarioParticipantKind,
    pub(crate) contract: String,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub(crate) enum ScenarioParticipantKind {
    App,
    Service,
    Device,
    Admin,
}

impl fmt::Display for ScenarioParticipantKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ScenarioParticipantKind::App => write!(f, "app"),
            ScenarioParticipantKind::Service => write!(f, "service"),
            ScenarioParticipantKind::Device => write!(f, "device"),
            ScenarioParticipantKind::Admin => write!(f, "admin"),
        }
    }
}

impl<'de> Deserialize<'de> for ScenarioParticipantKind {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        match s.as_str() {
            "app" => Ok(ScenarioParticipantKind::App),
            "service" => Ok(ScenarioParticipantKind::Service),
            "device" => Ok(ScenarioParticipantKind::Device),
            "admin" => Ok(ScenarioParticipantKind::Admin),
            other => Err(de::Error::custom(format!(
                "invalid participant kind: {other}, expected one of: app, service, device, admin"
            ))),
        }
    }
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct RawClientTestMatrix {
    cases: Vec<RawMatrixCase>,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct RawMatrixCase {
    id: String,
    fixture: String,
    title: String,
    coverage: Vec<String>,
    description: String,
    scenario: RawScenario,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct RawScenario {
    participants: Vec<RawScenarioParticipant>,
    given: Vec<String>,
    when: Vec<String>,
    then: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct RawScenarioParticipant {
    name: String,
    kind: ScenarioParticipantKind,
    contract: String,
}

pub(crate) fn load_client_test_matrix() -> Result<ClientTestMatrix, String> {
    let path = client_test_matrix_path()?;
    let contents = fs::read_to_string(&path)
        .map_err(|error| format!("failed to read {}: {error}", path.display()))?;
    let raw: RawClientTestMatrix = serde_json::from_str(&contents)
        .map_err(|error| format!("failed to parse {}: {error}", path.display()))?;
    validate_matrix(raw)
}

pub(crate) fn matrix_case_ids(matrix: &ClientTestMatrix) -> Vec<String> {
    let mut ids = matrix
        .cases
        .iter()
        .map(|case_entry| case_entry.id.clone())
        .collect::<Vec<_>>();
    ids.sort();
    ids
}

pub(crate) fn repo_root() -> Result<PathBuf, String> {
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    for ancestor in manifest_dir.ancestors() {
        if ancestor
            .join("integration/client-test-matrix.json")
            .exists()
            && ancestor.join("rust/Cargo.toml").exists()
            && ancestor.join("js/deno.json").exists()
        {
            return Ok(ancestor.to_path_buf());
        }
    }
    Err(format!(
        "failed to resolve repository root from {}",
        manifest_dir.display()
    ))
}

fn client_test_matrix_path() -> Result<PathBuf, String> {
    Ok(repo_root()?.join("integration/client-test-matrix.json"))
}

fn validate_matrix(raw: RawClientTestMatrix) -> Result<ClientTestMatrix, String> {
    let mut seen_ids = BTreeSet::new();
    let mut duplicate_ids = BTreeSet::new();
    let mut cases = Vec::with_capacity(raw.cases.len());
    let mut errors = Vec::new();

    for (index, raw_case) in raw.cases.into_iter().enumerate() {
        let context = format!("matrix case {}", index + 1);
        if !seen_ids.insert(raw_case.id.clone()) {
            duplicate_ids.insert(raw_case.id.clone());
        }
        validate_non_empty(&raw_case.id, &format!("{context} id"), &mut errors);
        validate_non_empty(
            &raw_case.fixture,
            &format!("{context} fixture"),
            &mut errors,
        );
        validate_non_empty(&raw_case.title, &format!("{context} title"), &mut errors);
        validate_non_empty(
            &raw_case.description,
            &format!("{context} description"),
            &mut errors,
        );
        for (coverage_index, coverage) in raw_case.coverage.iter().enumerate() {
            validate_non_empty(
                coverage,
                &format!("{context} coverage {}", coverage_index + 1),
                &mut errors,
            );
        }

        let scenario = parse_scenario(raw_case.scenario, &context, &mut errors);

        let expected_prefix = format!("{}.", raw_case.fixture);
        if !raw_case.id.starts_with(&expected_prefix) {
            errors.push(format!(
                "{context} id {} must start with fixture prefix {expected_prefix}",
                raw_case.id
            ));
        }
        cases.push(MatrixCase {
            id: raw_case.id,
            fixture: raw_case.fixture,
            title: raw_case.title,
            coverage: raw_case.coverage,
            description: raw_case.description,
            scenario,
        });
    }

    if !duplicate_ids.is_empty() {
        errors.push(format!(
            "client integration matrix has duplicate case ids: {}",
            duplicate_ids.into_iter().collect::<Vec<_>>().join(", ")
        ));
    }
    if !errors.is_empty() {
        return Err(errors.join("\n"));
    }

    Ok(ClientTestMatrix { cases })
}

fn parse_scenario(raw: RawScenario, context: &str, errors: &mut Vec<String>) -> Scenario {
    if raw.participants.is_empty() {
        errors.push(format!(
            "{context} scenario participants must be a non-empty array"
        ));
    }
    for (i, participant) in raw.participants.iter().enumerate() {
        let pctx = format!("{context} scenario participant {}", i + 1);
        validate_non_empty(&participant.name, &format!("{pctx} name"), errors);
        validate_non_empty(&participant.contract, &format!("{pctx} contract"), errors);
    }

    validate_non_empty_strings(&raw.given, &format!("{context} scenario given"), errors);
    validate_non_empty_strings(&raw.when, &format!("{context} scenario when"), errors);
    validate_non_empty_strings(&raw.then, &format!("{context} scenario then"), errors);

    let participants: Vec<ScenarioParticipant> = raw
        .participants
        .into_iter()
        .map(|p| ScenarioParticipant {
            name: p.name,
            kind: p.kind,
            contract: p.contract,
        })
        .collect();

    Scenario {
        participants,
        given: raw.given,
        when: raw.when,
        then: raw.then,
    }
}

fn validate_non_empty(value: &str, context: &str, errors: &mut Vec<String>) {
    if value.trim().is_empty() {
        errors.push(format!("{context} must be a non-empty string"));
    }
}

fn validate_non_empty_strings(values: &[String], context: &str, errors: &mut Vec<String>) {
    if values.is_empty() {
        errors.push(format!("{context} must be a non-empty array"));
    }
    for (i, entry) in values.iter().enumerate() {
        validate_non_empty(entry, &format!("{context} {}", i + 1), errors);
    }
}
